crates/ruff_linter/src/rules/ruff/rules/class_dict_annotations.rs

```diff
@@ -0,0 +1,131 @@
+/// for alternatives.
+///
+/// TLDR:
+///
+/// If you use a class with a custom metaclass and access `__annotations__`
+/// on the class, you may observe unexpected behavior; see
+/// [PEP 749](https://peps.python.org/pep-0749/#pep749-metaclasses) for some
+/// examples.
+///
+/// You can avoid these quirks by using `annotationlib.get_annotations()` on
+/// Python 3.14+, `inspect.get_annotations()` on Python 3.10+, or
+/// `typing_extensions.get_annotations(cls)` on Python < 3.10 with
+/// `typing_extensions` enabled.
+///
+/// ## Example
+///
+/// ```python
+/// cls.__dict__.get("__annotations__", {})
+/// ```
+///
+/// On Python 3.14+, use instead:
+/// ```python
+/// import annotationlib
+///
+/// annotationlib.get_annotations(cls)
+/// ```
+///
+/// On Python 3.10+, use instead:
+/// ```python
+/// import inspect
+///
+/// inspect.get_annotations(cls)
+/// ```
+///
+/// On Python < 3.10 with `typing_extensions` enabled, use instead:
+/// ```python
+/// import typing_extensions
+///
+/// typing_extensions.get_annotations(cls)
+/// ```
+///
+/// ## Fix safety
+///
+/// No autofix is currently provided for this rule.
+///
+/// ## Fix availability
+///
+/// No autofix is currently provided for this rule.
+///
+/// ## References
+/// - [Python Annotations Best Practices](https://docs.python.org/3.14/howto/annotations.html)
+#[derive(ViolationMetadata)]
+pub(crate) struct ClassDictAnnotations;
+
+impl Violation for ClassDictAnnotations {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Use `typing_extensions.get_annotations` (Python < 3.10 with `typing_extensions` enabled), `inspect.get_annotations` (Python 3.10+), or `annotationlib.get_annotations` (Python 3.14+) instead of `__dict__.get('__annotations__')`".to_string()
+    }
+}
+
+/// RUF061
+pub(crate) fn class_dict_annotations(checker: &Checker, call: &ExprCall) {
+    // Only apply this rule for Python 3.10 and newer unless `typing_extensions` is enabled.
+    if checker.target_version() < PythonVersion::PY310 && !checker.settings.typing_extensions {
+        return;
+    }
+
+    // Expected pattern: <identifier>.__dict__.get("__annotations__" [, <default>])
+    // Here, `call` is the `.get(...)` part.
+
+    // 1. Check that the `call.func` is `get`
+    let get_attribute = match call.func.as_ref() {
+        Expr::Attribute(attr) if attr.attr.as_str() == "get" => attr,
+        _ => return, // Not a call to an attribute named "get"
+    };
+
+    // 2. Check that the `get_attribute.value` is `__dict__`
+    let dict_attribute = match get_attribute.value.as_ref() {
+        Expr::Attribute(attr) if attr.attr.as_str() == "__dict__" => attr,
+        _ => return, // The object of ".get" is not an attribute named "__dict__"
+    };
+
+    // 3. Check that the `dict_attribute.value` is an Expr::Name (e.g., `cls`, `obj`)
```

Comment:
I don't think you need this check. There's no particular reason the class would be stored directly in a variable and not e.g. in a list.

Reply:
This check has been removed.